djpsa/halo/utils.py

```diff
@@ -1,43 +1,100 @@
+import time
 import redis
 import logging
 import requests
-
+from contextlib import contextmanager
 
 from django.conf import settings
 
 logger = logging.getLogger(__name__)
 
+ONE_HOUR = 3600
+REQUEST_TIMEOUT = 30
+LOCK_TIMEOUT = REQUEST_TIMEOUT + 1
+BLOCK_TIMEOUT = 60
 
-redis_client = redis.StrictRedis(
-        host=settings.REDIS['host'],
-        port=settings.REDIS['port'],
-        password=settings.REDIS['password'],
-    )
+redis_client = None
+
+
+def get_redis_client():
+    global redis_client
+
+    if not redis_client:
+        redis_client = redis.StrictRedis(
+            host=settings.REDIS['host'],
+            port=settings.REDIS['port'],
+            password=settings.REDIS['password'],
+        )
+
+    return redis_client
+
+
+class LockNotAcquiredError(Exception):
+    """Custom exception raised when the lock is not acquired."""
+    pass
+
+
+@contextmanager
+def redis_lock(lock_name):
+    lock = get_redis_client().lock(
+        lock_name, timeout=LOCK_TIMEOUT, blocking_timeout=BLOCK_TIMEOUT)
+    acquired = lock.acquire(blocking=True)
+    if not acquired:
+        raise LockNotAcquiredError()
+    try:
+        yield acquired
+    finally:
+        try:
+            lock.release()
+        except redis.exceptions.LockNotOwnedError:
+            pass
 
 
 def get_saved_token():
-    token = redis_client.get('halo_token')
-    if token:
+    token_data = get_redis_client().hgetall('halo_token')
+    if not token_data:
+        return None
+
+    token = token_data.get(b'token')
+    expiry = token_data.get(b'expiry')
+
+    # Check if the token has expired
+    if expiry and float(expiry) > time.time():
         return token.decode('utf-8')
 
+    # Token expired, remove it
+    rm_token()
+    return None
+
 
 def save_token(token):
     logger.debug('Saving token to Redis')
-    redis_client.set('halo_token', token)
+    # Subtract some time for buffer, say 60 seconds
+    expiry_time = time.time() + ONE_HOUR - 60
+    get_redis_client().hset('halo_token',
+                            mapping={'token': token, 'expiry': expiry_time})
```

Comment:
For expiry, just use Redis' own https://redis.io/docs/latest/commands/expire/

Reply:
Oh actually nevermind. As discussed- can you change this to use Django's own cache? https://docs.djangoproject.com/en/5.1/topics/cache/#accessing-the-cache, so users don't require Redis?

The cache has a timeout too so you still don't need to compare dates yourself.

Reply:
Sorry, for some reason I thought we discussed finishing this up then making an issue to convert it later.